### PR TITLE
feat: v1beta1 CRP validation for PickN Placement Policy

### PR DESCRIPTION
**Pull request by Arvindthiru**
### Description of your changes

Fixes #

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### CODE REVIEW

File: pkg/utils/validator/clusterresourceplacement.go [hunk 1]
```diff
@@ -141,32 +141,73 @@ func validatePolicyForPickAllPlacementType(policy *placementv1beta1.PlacementPol
...
+		}
 	}
-	return nil
+	return apiErrors.NewAggregate(allErr)
+}
+
+func validateTopologySpreadConstraints(topologyConstraints []placementv1beta1.TopologySpreadConstraint) error {
+	allErr := make([]error, 0)
+	for _, tc := range topologyConstraints {
+		if len(tc.TopologyKey) == 0 {
```
Comment (michaelawyu) [inline pkg/utils/validator/clusterresourceplacement.go:203]: Hi Arvind! `TopologyKey` is already a required field IIRC.

File: pkg/utils/validator/clusterresourceplacement.go [hunk 2]
```diff
@@ -141,32 +141,73 @@ func validatePolicyForPickAllPlacementType(policy *placementv1beta1.PlacementPol
...
 
-func validatePolicyForPickNPolicyType(_ *placementv1beta1.PlacementPolicy) error {
-	// TODO(Arvindthiru): implement this.
+func validatePolicyForPickNPolicyType(policy *placementv1beta1.PlacementPolicy) error {
 	allErr := make([]error, 0)
+	if len(policy.ClusterNames) > 0 {
+		allErr = append(allErr, fmt.Errorf("cluster names needs to be empty for policy type %s, only valid for PickFixed policy type", placementv1beta1.PickNPlacementType))
+	}
+	if policy.NumberOfClusters != nil {
+		if *policy.NumberOfClusters <= 0 {
```
Comment (michaelawyu) [inline pkg/utils/validator/clusterresourceplacement.go:161]: Hi Arvind! There is already a minimum = 0 constraint in the API I believe.

Comment (michaelawyu) [reply pkg/utils/validator/clusterresourceplacement.go:161]: Also, 0 is an allowed value.

Comment (michaelawyu) [reply pkg/utils/validator/clusterresourceplacement.go:161]: Some of the test cases might need an update accordingly.

File: pkg/utils/validator/clusterresourceplacement.go [hunk 3]
```diff
@@ -141,32 +141,70 @@ func validatePolicyForPickAllPlacementType(policy *placementv1beta1.PlacementPol
...
 
-func validatePolicyForPickNPolicyType(_ *placementv1beta1.PlacementPolicy) error {
-	// TODO(Arvindthiru): implement this.
+func validatePolicyForPickNPolicyType(policy *placementv1beta1.PlacementPolicy) error {
 	allErr := make([]error, 0)
+	if len(policy.ClusterNames) > 0 {
+		allErr = append(allErr, fmt.Errorf("cluster names needs to be empty for policy type %s, only valid for PickFixed policy type", placementv1beta1.PickNPlacementType))
+	}
+	if policy.NumberOfClusters != nil {
+		if *policy.NumberOfClusters < 0 {
```
Comment (ryanzhang-oss) [inline pkg/utils/validator/clusterresourceplacement.go:161]: is there a max?

File: pkg/utils/validator/clusterresourceplacement.go [hunk 4]
```diff
@@ -178,6 +216,15 @@ func validateClusterSelector(clusterSelector *placementv1beta1.ClusterSelector)
 	return apiErrors.NewAggregate(allErr)
 }
 
+func validatePreferredClusterSelectors(preferredClusterSelectors []placementv1beta1.PreferredClusterSelector) error {
+	allErr := make([]error, 0)
+	for _, preferredClusterSelector := range preferredClusterSelectors {
```
Comment (ryanzhang-oss) [inline pkg/utils/validator/clusterresourceplacement.go:221]: what  about weight?

File: pkg/utils/validator/clusterresourceplacement.go [hunk 5]
```diff
@@ -141,32 +141,70 @@ func validatePolicyForPickAllPlacementType(policy *placementv1beta1.PlacementPol
...
+		}
 	}
-	return nil
+	return apiErrors.NewAggregate(allErr)
+}
+
+func validateTopologySpreadConstraints(topologyConstraints []placementv1beta1.TopologySpreadConstraint) error {
+	allErr := make([]error, 0)
+	for _, tc := range topologyConstraints {
+		if len(tc.WhenUnsatisfiable) > 0 && tc.WhenUnsatisfiable != placementv1beta1.DoNotSchedule && tc.WhenUnsatisfiable != placementv1beta1.ScheduleAnyway {
```
Comment (ryanzhang-oss) [inline pkg/utils/validator/clusterresourceplacement.go:203]: what about MaxSkew?

File: pkg/utils/validator/clusterresourceplacement.go (hunk 4 above)
Comment (Arvindthiru) [reply pkg/utils/validator/clusterresourceplacement.go:221]: The API already has restrictions on weight range 
image

**Review by ryanzhang-oss — APPROVED** [review]

**Pull request merged**